text/0114-dynamic-component-configuration.md

```diff
@@ -0,0 +1,233 @@
+		}
+
+		// Some chart items, like a dimension filter card,
+
+		// might not support measures as inputs:
+
+		if (chart.supportsMeasures) {
+			props.measures = dashboardData.measures
+		}
+
+		// Some charts support colors in different ways:
+		if (chart.type === 'kpi' || chart.type === 'gauge') {
+			props.breakpoints = chart.customBreaks || this.defaultKpiBreaks;
+		} else if (chart.supportsColorPalette) {
+			props.colorPalette = this.colorPalette;
+		}
+		// Populate 'events'
+		if (chart.canFilter) {
+			eventHandlers.onfilter = this.handleFilter;
+		}
+		if (chart.canSort) {
+			eventHandlers.onsort = this.handleSort;
+		}
+		return {
+			ctor,
+			attributes,
+			properties: props,
+			events: eventHandlers
+		}
+
+	});
+
+}
+
+}
+```
+## Motivation
+
+Before this proposal, dynamic component creation was done using the `lwc:dynamic` directive. This directive required a component author to know component configuration details like the property and event names at design time which breaks down for use cases that need true dynamic component creation. For instance, in the case of property values or event handlers one would have to keep a reference to the element and set the information retrieved at runtime manually.
+
+```js
+element;
+
+@api
+moduleName;
+
+connectedCallback() {
+	const ctor = await import(this.moduleName);
+	this.element = createComponent('c-lazy-cmp', { is: ctor });
+	// element must be manually inserted into the dom
+	this.template.querySelector('div').appendChild(this.element)
+}
+
+@api
+
+set properties(newProperties) {
+
+this.props = newProperties;
+
+Object.assign(this.element, this.props);
+
+}
+
+get attributes() { return this.props; }
+
+```
+A new tool must be introduced that will allow component authors to configure their components with information retrieved at runtime without relying on maintaining element references or manually manipulating the DOM.
+
+## Proposal
+
+The proposal involves adding a new directive that captures the relevant information to configure a web component at runtime. The assumption is that this directive will be used in cases where the component module is unknown at runtime (therefore used alongside dynamic imports) and zero or more of the following are unknown at runtime: properties, attributes, or event handlers.
+
+The directive `lwc:bind` can retrieve component configuration data via a JavaScript object with a pre-defined shape. Constructor is the only required property in the configuration data.
+
+## General Invariants
+
+* It must not break the reactivity model or the general framework lifecycle
+
+* After the component is loaded, it should function within the lifecycle of the parent component
+
+* Changes to the configuration object must be reflected in the component (ie changing the properties will cause component properties to be set again)
+
+* Changing the constructor on a parent component will also trigger a cascade re-render of any child components 
```

Comment:
Changing a constructor requires unmounting the existing custom element and creating a new custom element. It actually doesn't re-render the children elements but creates brand new children elements.